src/node/http.ts

```diff
@@ -1,938 +1,253 @@
-   * domain (e.g. `coder.com` over `test.coder.com` if both are specified).
-   */
-  private getCookieDomain(host: string): string | undefined {
-    const idx = host.lastIndexOf(":")
-    host = idx !== -1 ? host.substring(0, idx) : host
-    if (
-      // Might be blank/missing, so there's nothing more to do.
-      !host ||
-      // IP addresses can't have subdomains so there's no value in setting the
-      // domain for them. Assume anything with a : is ipv6 (valid domain name
-      // characters are alphanumeric or dashes).
-      host.includes(":") ||
-      // Assume anything entirely numbers and dots is ipv4 (currently tlds
-      // cannot be entirely numbers).
-      !/[^0-9.]/.test(host) ||
-      // localhost subdomains don't seem to work at all (browser bug?).
-      host.endsWith(".localhost") ||
-      // It might be localhost (or an IP, see above) if it's a proxy and it
-      // isn't setting the host header to match the access domain.
-      host === "localhost"
-    ) {
-      logger.debug("no valid cookie doman", field("host", host))
-      return undefined
-    }
-
-    this.proxyDomains.forEach((domain) => {
-      if (host.endsWith(domain) && domain.length < host.length) {
-        host = domain
+    // Send the request off to be handled by Express.
+    ;(app as any).handle(req, res, () => {
+      if (!req._ws_handled) {
+        socket.destroy(new Error("Not found"))
       }
     })
+  })
+}
 
-    logger.debug("got cookie doman", field("host", host))
-    return host ? `Domain=${host}` : undefined
-  }
-
-  /**
-   * Return a response if the request should be proxied. Anything that ends in a
-   * proxy domain and has a *single* subdomain should be proxied. Anything else
-   * should return `undefined` and will be handled as normal.
-   *
-   * For example if `coder.com` is specified `8080.coder.com` will be proxied
-   * but `8080.test.coder.com` and `test.8080.coder.com` will not.
-   *
-   * Throw an error if proxying but the user isn't authenticated.
-   */
-  public maybeProxy(route: ProviderRoute, request: http.IncomingMessage): HttpResponse | undefined {
-    // Split into parts.
-    const host = request.headers.host || ""
-    const idx = host.indexOf(":")
-    const domain = idx !== -1 ? host.substring(0, idx) : host
-    const parts = domain.split(".")
-
-    // There must be an exact match.
-    const port = parts.shift()
-    const proxyDomain = parts.join(".")
-    if (!port || !this.proxyDomains.has(proxyDomain)) {
-      return undefined
-    }
-
-    // Must be authenticated to use the proxy.
-    route.provider.ensureAuthenticated(request)
-
-    return {
-      proxy: {
-        port,
-      },
-    }
+/**
+ * Patch Express routers to handle web sockets and async routes (since we have
+ * to patch `get` anyway).
+ *
+ * Not using express-ws since the ws-wrapped sockets don't work with the proxy
+ * and wildcards don't work correctly.
+ */
+function patchRouter(): void {
```

Comment:
I am sufficiently spooked by this whole thing, but I'm not sure I have the full context here. This all _seems_ like stuff that should be handleable through middlewares, and not actually mutating the underlying `Router` implementation. If this approach actually is necessary, is there any way for us to do this in a non-mutative way to `Router`? Can we extend the class / singleton and provide a new one to our express app? I feel confident there's got to be a better way here.

Reply:
I know we discussed on Slack but just so it's here as well this is the method that `express-ws` uses in case that affects our decision.

I'm not certain middleware can have the same effect because this patch will send websocket requests to `ws` handlers and everything else to the other handlers as per usual. With middleware we can modify the request but we can't change where it goes, I think? We also can't add a new `ws` method that way.

Some alternatives that came to mind as I was weaving this black magic:

  - Instead of a separate `ws` function every handler checks for `ws` on the `req` object and if it exists handle it appropriately (or error it if there's no websocket functionality at that route). My main objection with this is that the really clean registration interface Express provides gets a little muddled but it's not that bad.

  - Provide a wrapped router that everyone uses. `import { Router } from "express"` would be forbidden. This was actually the route I tried first but I aborted because I couldn't figure out a way to enforce it with code (I tried to override express.Router but it failed saying something about not having a getter) and it felt unfortunate that I (or a plugin) couldn't just import and use Express like usual. But I'm not totally opposed, ultimately it's not really that much more effort, just change the `from` on the import.

  - Have something completely separate. Like `import { wsRouter} from "./ws-router"` which has an Express-like syntax but maybe a bit simpler. I kind of like this one.

I'm happy to go with whatever we're most comfortable with!

Reply:
Well this sucks, this method doesn't work on Express 5.

Reply:
I got it working on Express 5 but I opted to use a separate app to handle web socket routes because there's middleware on the main app that doesn't make any sense to have on websocket routes, for example the cookie or body parser, the error html returned if routes throw an error, and so on. There's no jank or patching in this new version.